### PR TITLE
ref(types): Add types to rules

### PR DESCRIPTION
**Pull request by mgaeta**
Adding types to `src/sentry/rules/`.

### CODE REVIEW

**Review by evanh — COMMENTED** [review]
I think you should split this up, or at least separate the type annotations from the code restructuring (if possible). The issue is that I don't feel comfortable approving this since I don't know this code at all and there are changes that are not typing related.

File: src/sentry/rules/actions/base.py [hunk 1]
```diff
@@ -109,15 +131,9 @@ def _linked_issues(event, integration):
...
-    return _linked_issues(event, integration).values_list("key", flat=True)
-
-
-def has_linked_issue(event, integration):
-    return _linked_issues(event, integration).exists()
-
-
-def create_link(integration, installation, event, response):
+def create_link(
+    integration: Integration, installation: Any, event: Event, response: Response
```
Comment (evanh) [inline src/sentry/rules/actions/base.py:135]: The docs say that an `installation` is an `Installation` object. Is that not true? Why is `installation` typed as `Any`?

File: src/sentry/rules/actions/notify_event.py [hunk 2]
```diff
@@ -28,12 +30,12 @@ def get_plugins(self):
 
         return results
 
-    def after(self, event, state):
+    def after(self, event: Event, state: EventState) -> Any:
```
Comment (evanh) [inline src/sentry/rules/actions/notify_event.py:33]: There is a `Generator` type that you could use for the return type here. https://docs.python.org/3/library/typing.html#typing.Generator

File: src/sentry/integrations/pagerduty/notify_action.py [hunk 3]
```diff
@@ -74,7 +73,7 @@ def _validate_service(self, service_id: int, integration_id: int | None) -> None
                 params=params,
             )
 
-    def clean(self) -> Mapping[str, Any]:
+    def clean(self) -> Dict[str, Any]:
```
Comment (evanh) [inline src/sentry/integrations/pagerduty/notify_action.py:76]: This is more of a personal preference thing, but you can annotate this with `dict[str, Any]` and not have to import the typing class.

File: src/sentry/rules/actions/notify_event_sentry_app.py [hunk 4]
```diff
@@ -125,7 +128,7 @@ def self_validate(self) -> None:
                 f"Unexpected setting(s) '{extra_keys_string}' configured for {sentry_app.name}"
             )
 
-    def after(self, event: Event, state: str) -> Any:
+    def after(self, event: Event, state: EventState) -> Any:
```
Comment (evanh) [inline src/sentry/rules/actions/notify_event_sentry_app.py:131]: Same comment for Generator type.

File: src/sentry/rules/conditions/event_frequency.py [hunk 5]
```diff
@@ -49,20 +54,20 @@ class EventFrequencyForm(forms.Form):
...
     comparisonType = forms.ChoiceField(
-        choices=list(sorted(comparison_types.items(), key=lambda item: item[1])),
+        choices=list(sorted(comparison_types.items(), key=lambda item: item[1])),  # type: ignore
         required=False,
     )
     comparisonInterval = forms.ChoiceField(
         choices=[
             (key, label)
-            for key, (label, duration) in sorted(
+            for key, (label, duration_) in sorted(
```
Comment (evanh) [inline src/sentry/rules/conditions/event_frequency.py:63]: Why this change?

File: src/sentry/integrations/pagerduty/notify_action.py (hunk 3 above)
Comment (mgaeta) [reply src/sentry/integrations/pagerduty/notify_action.py:76]: I agree, I'll make a change.

File: src/sentry/rules/conditions/event_frequency.py (hunk 5 above)
Comment (mgaeta) [reply src/sentry/rules/conditions/event_frequency.py:63]: The variable is misbehaving because it's also defined inside `interval`. Since they're both unread I could rename both to `_`.

File: src/sentry/integrations/pagerduty/notify_action.py (hunk 3 above)
Comment (mgaeta) [reply src/sentry/integrations/pagerduty/notify_action.py:76]: Ugh I'm getting an error
```
Subscript for class "dict" will generate runtime exception; enclose type annotation in quotes
```
but it goes away when I do `Dict`.

Comment (evanh) [reply src/sentry/integrations/pagerduty/notify_action.py:76]: Weird. OK.

File: src/sentry/rules/conditions/event_frequency.py (hunk 5 above)
Comment (evanh) [reply src/sentry/rules/conditions/event_frequency.py:63]: Ah I see. Yeah if they are unused I would use `_`

**Review by evanh — APPROVED** [review]
These look good, pending fixing the typing CI.

File: src/sentry/rules/actions/base.py [hunk 6]
```diff
@@ -91,14 +115,19 @@ def get_integration(self):
             provider=self.provider,
         )
 
-    def get_installation(self):
+    def get_installation(self) -> Any:
```
Comment (wedamija) [inline src/sentry/rules/actions/base.py:118]: Should this return `IntegrationInstallation`?

File: src/sentry/rules/actions/services.py [hunk 7]
```diff
@@ -1,46 +1,53 @@
...
 
     @property
-    def service_type(self):
+    def service_type(self) -> str:
         return "legacy_plugin"
 
 
 class SentryAppService(PluginService):
-    def __init__(self, obj):
+    def __init__(self, obj: Any) -> None:
```
Comment (wedamija) [inline src/sentry/rules/actions/services.py:34]: Should this be `SentryApp`?

File: src/sentry/rules/conditions/event_frequency.py [hunk 8]
```diff
@@ -110,10 +113,11 @@ def passes(self, event, state):
         if not interval:
             return False
 
-        current_value = self.get_rate(event, interval, self.rule.environment_id)
+        # TODO(mgaeta): Bug: Rule is optional.
+        current_value = self.get_rate(event, interval, self.rule.environment_id)  # type: ignore
         return current_value > value
 
-    def query(self, event, start, end, environment_id):
+    def query(self, event: Event, start: datetime, end: datetime, environment_id: str) -> Any:
```
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:120]: This should return `int`, since we just cast it as an int in the only place that uses it.

File: src/sentry/rules/conditions/event_frequency.py [hunk 9]
```diff
@@ -124,14 +128,14 @@ def query(self, event, start, end, environment_id):
         )
         return query_result
 
-    def query_hook(self, event, start, end, environment_id):
+    def query_hook(self, event: Event, start: datetime, end: datetime, environment_id: str) -> Any:
```
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:131]: This should also return `int`

File: src/sentry/rules/conditions/event_frequency.py [hunk 10]
```diff
@@ -110,10 +113,11 @@ def passes(self, event, state):
         if not interval:
             return False
 
-        current_value = self.get_rate(event, interval, self.rule.environment_id)
+        # TODO(mgaeta): Bug: Rule is optional.
```
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:116]: It probably shouldn't be optional on the base class, I think we will always instantiate this with a `Rule`

File: src/sentry/rules/conditions/event_frequency.py [hunk 11]
```diff
@@ -151,21 +155,26 @@ def get_rate(self, event, interval, environment_id):
...
+        delta = abs(self.rule.date_added - self.project.date_added)  # type: ignore
+        guess: bool = delta.total_seconds() < 30 and self.rule.label == DEFAULT_RULE_LABEL  # type: ignore
+        return guess
 
 
 class EventFrequencyCondition(BaseEventFrequencyCondition):
     label = "The issue is seen more than {value} times in {interval}"
 
-    def query_hook(self, event, start, end, environment_id):
+    def query_hook(self, event: Event, start: datetime, end: datetime, environment_id: str) -> Any:
```
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:177]: Should return `int`

File: src/sentry/rules/conditions/event_frequency.py [hunk 12]
```diff
@@ -252,13 +262,13 @@ def __init__(self, *args, **kwargs):
...
 
-    def query_hook(self, event, start, end, environment_id):
+    def query_hook(self, event: Event, start: datetime, end: datetime, environment_id: str) -> int:
         project_id = event.project_id
         cache_key = f"r.c.spc:{project_id}-{environment_id}"
         session_count_last_hour = cache.get(cache_key)
         if session_count_last_hour is None:
             with options_override({"consistent": False}):
-                session_count_last_hour = release_health.get_project_sessions_count(
+                session_count_last_hour = release_health.get_project_sessions_count(  # type: ignore
```
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:271]: Why ignore here? Could we just cast as `int`?

Comment (mgaeta) [reply src/sentry/rules/conditions/event_frequency.py:271]: The problem isn't with `session_count_last_hour` it's that mypy can't find `get_project_sessions_count()` on `release_health`.
